stdlib/public/core/Arrays.swift.gyb

```diff
@@ -2235,6 +2235,25 @@ extension ${Self} : Equatable where Element : Equatable {
   }
 }
 
+extension ${Self} : Hashable where Element : Hashable {
+  /// The hash value for the array.
+  ///
+  /// Two arrays that are equal will always have equal hash values.
+  ///
+  /// Hash values are not guaranteed to be equal across different executions of
+  /// your program. Do not save hash values to use during a future execution.
+  @_inlineable // FIXME(sil-serialize-all)
+  public var hashValue: Int {
+    // FIXME(ABI)#177: <rdar://problem/18915294> Issue applies to Array too
+    var result: Int = 0
+    for element in self {
+      result = _combineHashValues(result, element.hashValue)
+    }
+    result = _mixInt(result)
```

Comment:
Call to `_mixInt` here is superfluous -- `Set` and `Dictionary` already calls it on every hash returned from `hashValue`. (It is probably okay to call it twice, but there doesn't seem to be a good reason to do so.)

Reply:
I recommended that @mortenbekditlevsen model these after synthesized conformances, which do call a final `_mixInt`. Part of the motivation here is that `_combineHashValues` is deterministic, but `_mixInt` can (eventually) have per-execution variation based on the magic number. I think if we're to abandon the final `_mixInt`, we ought to consider it wholesale and include the synthesized implementations in the discussion.

Reply:
The final `_mixInt` is done by `Set` and `Dictionary`; there is no need to do it twice.

Reply:
Sorry, I’m not sure I understand your comment.

We’re still discussing `Array` conformance, right? There are no other calls to `_mixInt` here—am I missing something?

Reply:
Calling `_mixInt` directly before returning from `hashValue` is unnecessary, because `Set` and `Dictionary` will call it for you (in [`_squeezeHashValue`][shv], which standard hashed collections use to generate bucket indices).

Doing it anyway does not improve distribution, it just slows things down. As far as I know, `_mixInt` wasn't designed to be used in the implementation of `hashValue` at all.

[shv]: https://github.com/apple/swift/blob/master/stdlib/public/core/Hashing.swift#L184

Reply:
Ah...got it.